espflash/src/targets/mod.rs

```diff
@@ -304,16 +308,57 @@ pub trait ReadEFuse {
     /// Returns the base address of the eFuse register
     fn efuse_reg(&self) -> u32;
 
+    /// Returns the offset of BLOCK0 relative to the eFuse base register address
+    fn block0_offset(&self) -> u32;
+
+    /// Returns the size of the specified block for the implementing target
+    /// device
+    fn block_size(&self, block: usize) -> u32;
+
+    /// Given an active connection, read the specified field of the eFuse region
+    #[cfg(feature = "serialport")]
+    fn read_efuse(&self, connection: &mut Connection, field: EfuseField) -> Result<u32, Error> {
+        let mask = if field.bit_count == 32 {
+            u32::MAX
+        } else {
+            (1u32 << field.bit_count) - 1
+        };
+
+        let shift = field.bit_start % 32;
+
+        let value = self.read_efuse_raw(connection, field.block, field.word)?;
+        let value = (value >> shift) & mask;
+
+        Ok(value)
+    }
+
+    /// Read the raw word in the specified eFuse block, without performing any
+    /// bit-shifting or masking of the read value
     #[cfg(feature = "serialport")]
-    /// Given an active connection, read the nth word of the eFuse region
-    fn read_efuse(&self, connection: &mut Connection, n: u32) -> Result<u32, Error> {
-        let reg = self.efuse_reg() + (n * 0x4);
-        connection.read_reg(reg)
+    fn read_efuse_raw(
+        &self,
+        connection: &mut Connection,
+        block: u32,
+        word: u32,
+    ) -> Result<u32, Error> {
+        let block0_addr = self.efuse_reg() + self.block0_offset();
+        let block_offset = if block == 0 {
+            0
+        } else {
+            self.block_size(block as usize - 1)
+        };
```

Comment:
This isn't correct. We shouldn't offset by the size of the previous block, we should offset by the total size of all previous blocks. Something like this works:

```rust
        let mut block_offset = 0;
        for b in 0..block {
            block_offset += self.block_size(b as usize);
        }
```

```
Chip type:         esp32s2 (revision v1.0)
Crystal frequency: 40 MHz
Flash size:        8MB
Features:          WiFi, No Embedded Flash, Embedded PSRAM 2MB, ADC and temperature sensor calibration in BLK2 of efuse V2
```